lib/makara/context.rb

```diff
@@ -1,83 +1,144 @@
-        previously_sticky.fetch(config_id) do
-          stuck?(Makara::Context.get_previous, config_id)
+      def commit(headers, cookie_options = {})
+        current.release_expired
+        if current.dirty?
+          cookie = DEFAULT_OPTIONS.merge(cookie_options)
+          Rack::Utils.set_cookie_header! headers, IDENTIFIER, cookie.merge(current.to_cookie_options)
         end
       end
 
-      # Called by `Proxy#stick_to_master!` to stick subsequent requests to
-      # master. They'll see the current context as their previous context
-      # when they're asking whether they should be stuck to master.
-      def stick(context, config_id, ttl)
-        Makara::Cache.write(cache_key_for(context, config_id), '1', ttl)
+      def release(config_id)
+        current.release(config_id)
       end
 
-      def stuck?(context, config_id)
-        !!Makara::Cache.read(cache_key_for(context, config_id))
+      def release_all
+        current.release_all
       end
 
       protected
-
-      def previously_sticky
-        fetch(:makara_previously_sticky) { Hash.new }
-      end
-
-      def cache_key_for(context, config_id)
-        "makara::#{context}-#{config_id}"
+      def current
+        fetch(:makara_current_context) { new({}) }
       end
 
       def fetch(key)
-        get(key) || set(key,yield)
+        get(key) || set(key, yield)
       end
 
       if Thread.current.respond_to?(:thread_variable_get)
         def get(key)
           Thread.current.thread_variable_get(key)
         end
 
-        def set(key,value)
-          Thread.current.thread_variable_set(key,value)
+        def set(key, value)
+          Thread.current.thread_variable_set(key, value)
         end
       else
         def get(key)
           Thread.current[key]
         end
 
-        def set(key,value)
+        def set(key, value)
           Thread.current[key]=value
         end
       end
 
+      private
+
+      # Pairs of {config_id}:{timestamp}, separated by "|"
+      # config_id1:1518270031.3132212|config_id2:1518270030.313232 ..
+      def parse(cookie_string)
+        return {} if cookie_string.empty?
+
+        states = cookie_string.split("|")
+        states.each_with_object({}) do |state, data|
+          config_id, timestamp = state.split(":")
+          data[config_id] = timestamp.to_f if config_id && timestamp
+        end
+      end
+    end
+
+    attr_accessor :data
+
+    def initialize(data)
+      @data = data
+      @dirty = false
+    end
+
+    def stick(config_id, ttl)
+      data[config_id] = Time.now.to_f + ttl.to_f
+      @dirty = true
+    end
+
+    def stuck?(config_id)
+      data[config_id] && !expired?(data[config_id])
+    end
+
+    # Indicates whether there have been changes to the context that need
+    # to be persisted when the request finishes
+    def dirty?
+      @dirty
+    end
+
+    def release(config_id)
+      @dirty ||= !!data.delete(config_id)
+    end
+
+    def release_expired
```

Comment:
Hmm... do you mean `config_id`s that don't correspond to any existing configuration? That is, if a configuration is changed for example, so the id changes? 
Yes, I thought in that case they would only live in the cookie until their expiration time passes, which should be very short in general. They'll be removed by this method when that happens.

Reply:
Yeah, that's what I meant. If someone were to stop using Makara for redis, would it be in the cookie forever. As it is now, it wold go away when the TTL is over, so that's seems fine as these should be fairly short. That small risk seems better than accidentally removing it in some transition situation. So carry on.